polyfills.js

```diff
@@ -80,38 +81,120 @@ function patch (fs) {
+      } catch (e) {
+        if (e.code === "ENOTEMPTY") {
+          // target directory not empty, can't rename
+          if (cb) cb(e)
+          return
+        }
+        // ignore other errors
+      }
       var start = Date.now()
-      var backoff = 0;
+      var backoff = 0
+      var backoffUntil = start + win32MaxBackoff
       fs$rename(from, to, function CB (er) {
-        if (er
-            && (er.code === "EACCES" || er.code === "EPERM")
-            && Date.now() - start < 60000) {
+        if (er && (er.code === "EACCES" || er.code === "EPERM") && Date.now() < backoffUntil) {
           setTimeout(function() {
-            fs.stat(to, function (stater, st) {
-              if (stater && stater.code === "ENOENT")
-                fs$rename(from, to, CB);
-              else
-                cb(er)
+            fs.stat(from, function (erFrom, statFrom) {
+              fs.stat(to, function (erTo, statTo) {
+                if (erFrom && !erTo) {
+                  // If the source no longer exists we 
+                  // can probably assume it was moved
+                  if (cb) cb(null)
+                } else if (
+                    statFrom && statTo &&
+                    statFrom.size === statTo.size && 
+                    statFrom.ctime === statTo.ctime
+                  ) {
+                  // If the source and target have 
+                  // the same size and ctime, we
+                  // can assume it was moved
+                  if (cb) cb(null)
+                } else
+                  fs$rename(from, to, CB)
+              })
             })
           }, backoff)
-          if (backoff < 100)
-            backoff += 10;
-          return;
+          if (backoff < 250)
+            backoff += 10
+        } else if (backoff && er && er.code === "ENOENT") {
+          // The source does no longer exist so we
+          // can assume it was moved during one of the tries
+          if (cb) cb(null)
+        } else {
+          if (cb) cb(er)
         }
-        if (cb) cb(er)
       })
     }})(fs.rename)
+
+    fs.renameSync = (function (fs$renameSync) { return function (from, to) {
+      try {
+        var stat = fs.statSync(to)
+        if (!stat) return
+        if (stat.isDirectory()) {
+          fs.rmdirSync(to)
+        } else {
+          fs.unlinkSync(to)
+        }
+      } catch (e) {
+        if (e.code === "ENOTEMPTY") {
+          // target directory not empty, can't rename
+          throw e
+        }
+        // ignore other errors
+      }
+      var start = Date.now()
+      var backoff = 0
+      var backoffUntil = start + win32MaxBackoff
+      function tryRename () {
+        try {
+          fs$renameSync(from, to)
+        } catch (e) {
+          if ((e.code === "EACCS" || e.code === "EPERM") && start < backoffUntil) {
+            if (backoff < 100)
+              backoff += 10
+            var waitUntil = Date.now() + backoff
+            while (waitUntil > Date.now()){}
```

Comment:
This is extremely hazardous, as it has the potential to lock up the event loop for up to 5 seconds.

It'd probably be safer to just let synchronous renames fail.